### PR TITLE
➕ Add FastFlow Model

### LINKED ISSUE #261: Add FastFlow Model
### Discussed in https://github.com/openvinotoolkit/anomalib/discussions/189

<div type='discussions-op-text'>

<sup>Originally posted by **innat** February 23, 2022</sup>
**Is your feature request related to a problem? Please describe.**

  - New sota: https://paperswithcode.com/sota/anomaly-detection-on-mvtec-ad
  - paper: https://arxiv.org/pdf/2111.07677v2.pdf
</div>

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- Add fastflow implementation

- Fixes #261 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/fastflow/__init__.py
Comment (djdameln) [inline anomalib/models/fastflow/__init__.py:10]: Do we want to expose all these classes here in the init? For the other models we only list the `Lightning` class here. I'm fine with either option, but at least we should be consistent.

File: anomalib/models/fastflow/config.yaml
Comment (djdameln) [inline anomalib/models/fastflow/config.yaml:10]: `num_workers` is usually set to a higher default value, for the other models it's 36.

File: anomalib/models/fastflow/torch_model.py
Comment (djdameln) [inline anomalib/models/fastflow/torch_model.py:198]: Does pylint allow single letter variable names?

File: anomalib/models/fastflow/torch_model.py
Comment (djdameln) [inline anomalib/models/fastflow/torch_model.py:181]: This method is a bit hard to read. Maybe you could split it into several sub-methods. Something like:
```
if isinstance(self.feature_extractor, VisionTransformer):
    features = self.get_transformer_features(x)
elif isinstance(self.feature_extractor, Cait):
    features = self.get_cait_features(x)
else:
    features = self.feature_extractor(x)
    
hidden_variables = self.get_hidden_variables(features)
```

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks! I have a few comments, mainly minor

File: anomalib/models/components/base/anomaly_module.py
Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:178]: fastflow has long training time. without the performance being logged to the progress bar, it's hard to follow what's happening. 

pixel-metrics should be logged by default. However, if it's not available, image-metrics should be logged.

File: anomalib/models/fastflow/__init__.py
Comment (samet-akcay) [reply anomalib/models/fastflow/__init__.py:10]: I think it would be good to be able to import them. Especially when we use them as components. We could do the same for the other algos

File: anomalib/models/fastflow/config.yaml
Comment (samet-akcay) [reply anomalib/models/fastflow/config.yaml:10]: thanks for spotting this

File: anomalib/models/fastflow/torch_model.py
Comment (samet-akcay) [reply anomalib/models/fastflow/torch_model.py:198]: I temporarily ignore that in L192. As I said, this `FastFlowModel` would require some refactor later.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks! Only a few minor comments

File: anomalib/models/components/base/anomaly_module.py
Comment (ashwinvaidya17) [reply anomalib/models/components/base/anomaly_module.py:178]: I am clear. Maybe we can wait for @djdameln

**Review by ashwinvaidya17 — APPROVED** [review]

File: tests/nightly/models/test_model_nightly.py
Comment (ashwinvaidya17) [inline tests/nightly/models/test_model_nightly.py:48]: I don't think this will work as the performance thresholds for fastflow haven't been updated in the performance threshold file.

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]

File: anomalib/models/fastflow/README.md
Comment (ashwinvaidya17) [inline anomalib/models/fastflow/README.md:51]: Can you add the average score? Also, a minor issue but this format is a bit different from our other models (rows <-> columns). Maybe for consistency we should decide on a single format. Maybe this can be addressed in a different PR.

Comment (samet-akcay) [reply anomalib/models/fastflow/README.md:51]: We could perhaps modify all the tables once we have all the benchmark results merged

**Pull request merged**
